File: 152_maximum_product_subarray/maximum_product_subarray.go
package main

import "fmt"

type Solution struct{}

func (s *Solution) maximumProductSubarray(nums []int) int {
	maxProd, minProd, result := nums[0], nums[0], nums[0]

	for i := 1; i < len(nums); i++ {
		num := nums[i]
		if num < 0 {
			maxProd, minProd = minProd, maxProd
		}

		maxProd = max(num, num*maxProd)
		minProd = min(num, num*minProd)
		result = max(result, maxProd)
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	nums := []int{2, 3, -2, 4}
	s := Solution{}
	fmt.Println(s.maximumProductSubarray(nums))
}

File: 152_maximum_product_subarray/maximum_product_subarray.py
from typing import List

class Solution:
    def maximum_product_subarray(self, nums: List[int]) -> int:
        max_prod = min_prod = result = nums[0]

        for num in nums[1:]:
            if num < 0:
                max_prod, min_prod = min_prod, max_prod

            max_prod = max(num, num * max_prod)
            min_prod = min(num, num * min_prod)
            result = max(result, max_prod)

        return result


if __name__ == "__main__":
    nums = [2, 3, -2, 4]
    solution = Solution()
    print(solution.maximum_product_subarray(nums))
